logging.cpp

```diff
@@ -0,0 +1,89 @@
+#include "logging.h"
+
+#include <sys/stat.h>
+#include <stdarg.h>
+
+using namespace std;
+
+Log::LogMsg::LogMsg(std::string& msg){
+  this->message = msg;
+}
+
+ostream& operator<<(ostream& logstream, const Log::LogMsg& msg){
+  return logstream << msg.message;
+}
+
+Log::Log(): msg_queue(), trans_mutex(), run_cv(){
+  is_open = false;
+}
+
+Log& Log::init(){
+  static Log instance;
+  return instance;
+}
+
+void Log::thread_handle(){
+  while(1){
+    unique_lock<mutex> flck(this->trans_mutex);
+    while(msg_queue.empty()){
+      if (!is_open){
+        return;
+      }
+      this->run_cv.wait(flck);
```

Comment:
Non-blocking: Add a comment pointing out that this automatically releases and later re-acquires `flck`.